core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -0,0 +1,257 @@
+      context.fail(BAD_REQUEST, e);
+      return;
+    }
+
+    // load existing keys
+    final Set<String> existingPubkeys =
+        artifactSignerProvider.availableIdentifiers().stream()
+            .map(IdentifierUtils::normaliseIdentifier)
+            .collect(Collectors.toSet());
+
+    // filter out already loaded keys for slashing data import
+    final List<String> nonLoadedPubkeys =
+        pubkeysToImport.stream()
+            .filter(key -> !existingPubkeys.contains(key))
+            .collect(Collectors.toList());
+
+    // read slashing protection data if present and import data matching non-loaded keys to import
+    // only
+    if (slashingProtection.isPresent()
+        && !StringUtils.isEmpty(parsedBody.getSlashingProtection())) {
+      try {
+        final InputStream slashingProtectionData =
+            new ByteArrayInputStream(
+                parsedBody.getSlashingProtection().getBytes(StandardCharsets.UTF_8));
+        slashingProtection.get().importDataWithFilter(slashingProtectionData, nonLoadedPubkeys);
+      } catch (Exception e) {
+        context.fail(BAD_REQUEST, e);
+        return;
+      }
+    }
+
+    final List<ImportKeystoreResult> results = new ArrayList<>();
+    for (int i = 0; i < parsedBody.getKeystores().size(); i++) {
+      final String pubkey = pubkeysToImport.get(i);
+      try {
+        final String jsonKeystoreData = parsedBody.getKeystores().get(i);
+        final String password = parsedBody.getPasswords().get(i);
+
+        if (existingPubkeys.contains(pubkey)) {
+          // keystore already loaded
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.DUPLICATE, null));
+        } else {
+          // new keystore to import
+          // 1. validate and decrypt the keystore
+          final BlsArtifactSigner signer =
+              decryptKeystoreAndCreateSigner(jsonKeystoreData, password);
+          // 2. write keystore file to disk
+          createKeyStoreYamlFileAt(pubkey, jsonKeystoreData, password);
+          // 3. add the new signer to the provider to make it available for signing
+          artifactSignerProvider.addSigner(signer).get();
+          // 4. finally, add result to API response
+          results.add(new ImportKeystoreResult(ImportKeystoreStatus.IMPORTED, null));
+        }
+      } catch (Exception e) {
+        // cleanup the current key being processed and continue
+        removeSignersAndCleanupImportedKeystoreFiles(List.of(pubkey));
+        results.add(
+            new ImportKeystoreResult(
+                ImportKeystoreStatus.ERROR, "Error importing keystore: " + e.getMessage()));
+      }
+    }
+
+    try {
+      context
+          .response()
+          .putHeader(CONTENT_TYPE, JSON_UTF_8)
+          .setStatusCode(SUCCESS)
+          .end(objectMapper.writeValueAsString(new ImportKeystoresResponse(results)));
+    } catch (Exception e) {
+      removeSignersAndCleanupImportedKeystoreFiles(pubkeysToImport);
```

Comment:
Should this be `nonLoadedPubkeys` rather than `pubkeysToImport`?
In the case of a duplicate, we wouldn't want to remove the original would we?

Reply:
true, should be safer. technically this won't fail since we control JSON serialization which is the only thing that throws here. But definitely safer, I'll make the change